include/envoy/matcher/matcher.h

```diff
@@ -0,0 +1,174 @@
+#pragma once
+
+#include <memory>
+#include <string>
+
+#include "envoy/common/pure.h"
+#include "envoy/config/common/matcher/v3/matcher.pb.h"
+#include "envoy/config/typed_config.h"
+
+#include "absl/strings/string_view.h"
+#include "absl/types/optional.h"
+#include "external/envoy_api/envoy/config/core/v3/extension.pb.h"
+
+namespace Envoy {
+namespace Matcher {
+
+// This file describes a MatchTree<DataType>, which traverses a tree of matches until it
+// either matches (resulting in either an action or a new tree to traverse) or doesn't match.
+// The matching might stop early if either the data is not available at all yet, or if more data
+// might result in a match.
+
+// By returning a new tree when an OnMatch results in a new tree, matching can be resumed from
+// this tree should more data be required to complete matching. This avoids having to start
+// from the beginning every time. At some point we might support resuming for any node in the match
+// tree: this requires some careful handling of tracking which on_no_match to use should we fail to
+// match.
+//
+// All the matching is performed on strings: a DataInput<DataType> is used to extract a specific
+// string from an instance of DataType, while an InputMatcher is used to determine whether the
+// extracted string is a match.
+//
+// For example, DataType might be the type HttpDataInput, allowing
+// for the use of HttpRequestHeaders : DataInput<HttpDataInput>, which is configured with the name
+// of the header to extract from the request headers.
+//
+// In cases where the data to match on becomes available over time, this would be fed into the
+// DataType over time, allowing matching to be re-attempted as more data is made available. As such
+// whenever we extract data from a DataInput, we make note of whether the data might change and
+// pause matching until we either match or have all the data. It would then fall on the caller to
+// both provide more information to the DataType and to resume matching.
+template <class DataType> class MatchTree;
+
+template <class DataType> using MatchTreeSharedPtr = std::shared_ptr<MatchTree<DataType>>;
+
+/**
+ * Action provides the interface for actions to perform when a match occurs. It provides no
+ * functions, as implementors are expected to downcast this to a more specific action.
```

Comment:
Sorry can you provide a more concrete example? I'm a little confused how this is actually going to be used. Who is going to downcast it and how will they know what to downcast it to? I'm a little confused how this is going to work in terms of having a mostly constant match tree, but then allow actions to be performed on some type of concrete instance. Does there need to be some type of `invoke()` function that takes some type of param? I'm still confused on exactly how that would be wired up.

/wait-any

Reply:
My thinking was that for the HCM we would have something like this 

```
struct SkipFilterAction : Action {};
```

```
struct PerformTapAction : Action {};
```

When the HCM gets a match result, it would attempt to cast the action to the types handled by the HCM. If this fails, it would propagate the action to the associated filter. 

While not part of this PR, I think we'll end up having a validator that inspects the tree as it is created, ensuring that the tree is compatible with the filter or whatever other context we're creating a match tree for. This mechanism would then allow the HCM to know what types it can handle (and attempt to downcast to all of them) before passing it along to the filter (which should handle the remaining ones). 

Open to suggestions here, the idea of having to iterate over a bunch of types in the HCM to look for the right type to downcast isn't the most elegant.

Reply:
It's only marginally better than using RTTI, but here is what I suggest:
1) Add a `typeUrl()` method which returns a string.
2) Every action is a registered type (even the built-in ones). I think this allows for HCM actions, filter actions, etc.
3) Return some type of opaque templated optional object which depends on the type_url and requires a dynamic_cast.

I think this is a slightly better use of RTTI and more inline with what we do elsewhere? WDYT?

Reply:
Note I would do (3) as an accessor method so we can check casts with ASSERTs, etc., but not strictly required.

Reply:
I pushed a change for this, lmk if this is along what you were thinking. It wasn't clear how I'd use the `typeUrl` with the cast helper, so I just assumed it would fall on the caller to figure out the right cast